### PR TITLE
Add a new event for process exiting

### PR DESCRIPTION
**Pull request by mgood**
Send a new event when a container process exits without being requested. This
event can be used when a program needs to detect that a container exited
unexpectedly.

This event is sent after "die" when the exit was not expected for another
command like "stop", "kill", or "restart".

This addresses part of #10654 regarding how to distinguish unexpected exits, though the other part of that issue seems to be clarifying the documentation on the "die" event.

This event is currently called "process-exit" but I'm not attached to that name and it can be easily changed. Maybe just "exited" would be sufficient, since this is similar to the distinction of [`wait()`](http://man7.org/linux/man-pages/man2/wait.2.html) return codes, where `WIFEXITED(status)` is true if the process ended itself via `exit()` vs. `WIFSIGNALED(status)` when the process was ended via a signal.

Closes #9214

### CODE REVIEW

**Comment by mgood** [comment]
Hmm, it looks like the build result for "janky" above is from my old PR #11826 for the same commit, but before I renamed the branch.

This was the Jenkins job triggered by this PR, which passed:
https://jenkins.dockerproject.com/job/Docker-PRs/4350/console

**Comment by jfrazelle** [comment]
i triggered a rebuild

**Comment by progrium** [comment]
`exit` sounds right. Though I think in an ideal would, "die" becomes "exit" and this becomes "die".

**Comment by mgood** [comment]
Just rebased & renamed event to just "exit".

**Comment by icecrime** [comment]
Sorry, I'm not really convinced by it being a *new* type of event :-( My understanding is that I'm first notified for `die` which means "the container exited", and *then* by `exit` which means "oh, that previous `die` event? totally unsolicited". I agree it's not as easy, but I think I'd rather see the event payload hold some context along with the action rather than a new type of event.

**Comment by progrium** [comment]
True. It's not ideal, but I think it follows the pattern that has already
been established by other die related events.

On Fri, Apr 3, 2015 at 10:37 AM, Arnaud Porterie <notifications@github.com>
wrote:

> Sorry, I'm not really convinced by it being a *new* type of event :-( My
> understanding is that I'm first notified for die which means "the
> container exited", and *then* by exit which means "oh, that previous die
> event? totally unsolicited". I agree it's not as easy, but I think I'd
> rather see the event payload hold some context along with the action rather
> than a new type of event.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/11827#issuecomment-89327333>.
>



-- 
Jeff Lindsay
http://progrium.com

**Comment by mgood** [comment]
@icecrime have there been other proposals for richer event data? It does sound interesting, but as @progrium mentioned, I was trying to follow the existing pattern with this PR.

When a container is stopped or killed, you get "die, stop" or "die, kill" as the sequence of events, so the event that triggered it to die comes second. So for consistency with those events, I included the "exit" event after "die", since it was the process calling "exit()" that triggered the container to die.

I've also been working on a diagram of the container events to help better document this information for users of the events API:
Docker events flow

If the events API was extended to provide event-specific metadata, do you have ideas on how that would be used for other events? If it's just for "die", do you think the added complexity is worthwhile?

**Comment by cpuguy83** [comment]
Not sure how this is much different than tracking whether the `die` event was preceeded by a `stop`/`restart`

**Comment by mgood** [comment]
@cpuguy83 this would be a little easier to work around if the `stop` event preceded `die`, but the events are emitted the other way around (see the diagram above for reference).

Yes, if `stop` and `kill` came first, I could keep a buffer of those events, so that when I receive a `die` event, I can check what preceded it. However, since they are emitted in the opposite order, I would have to wait after I see a `die` event to check whether it's followed by another event. I could set a timer, and assume that after some small timeout, if I hadn't received a `stop` or `kill` event for the container, that the `die` was unsolicited. However, this would be inefficient, and could be flakey if there is a delay in the event stream.

Changing the order now may be surprising, and backwards-incompatible for some users, so I'm not advocating that, but some deterministic means of knowing when a container exited without an explicit request would be helpful.

**Comment by LK4D4** [comment]
design is ok for me. It was really something that we missed.
ping @docker/core-maintainers

**Comment by duglin** [comment]
When I run: `docker run -ti ubuntu true` I see the following events:
```
create
start
die
exit
```
This doesn't feel right.  If `die` means that the container stopped, I agree with @icecrime that I would prefer some additional metadata on the `die` to indicate why it stopped.  e.g.  exit code=0 etc...  If we want to include additional info, like Docker sent it a signal, then I would prefer to see a dedicated event for that (e.g.  `signal(9)`) or something since I view the sending of the various signals as separate actions taken on the container - and there could be more than one - which could be important for tracking/debugging purposes.

```
create
start
die (exit status: 0)
```
Seems to make sense to me

**Comment by icecrime** [comment]
@mgood Sorry for the lack of attention on this PR. I stick to what I said earlier, but at the same time we don't have anything better, and the feature is indeed useful.

So +1 on design, do you please mind rebasing so we can go with code review? Thanks!

**Comment by duglin** [comment]
@mgood can you summarize what the events would look like for the various cases? Just a list like:
```
normal: create, start, die
exit normal:...
exit bad rc: ...
exit via kill/signal: ...
```
so we can compare and make sure it looks right

**Comment by icecrime** [comment]
@mgood Are you available to continue on this? Any maintainer up for carrying? If not I'll close it in a few days.

**Comment by cpuguy83** [comment]
@icecrime So we're good with design I assume?

**Comment by mgood** [comment]
@icecrime yeah, sorry let me catch back up on this. It shouldn't take too long to fix the merge conflicts once I've got it a little more fresh in my mind.

**Comment by icecrime** [comment]
@cpuguy83 Yes sorry, moving to code review.
@mgood Thanks, just ping us there we'll try to be more reactive with the code review.

**Comment by mgood** [comment]
@icecrime thanks, the latest push resolves the conflicts. I ran the affected events tests locally which passed, but I'll keep an eye out for the full CI results.

**Comment by mgood** [comment]
Spoke too soon :-/ looks like the tests I added for "kill" and "stop" failed in CI. I'll check on those. Looks like I may need to change my test command from what I used before to filter the tests to run.

**Comment by mgood** [comment]
Ah, ok found the new flags for go-check to filter the tests. I'll look into it.

File: daemon/container.go [hunk 1]
```diff
@@ -253,6 +253,7 @@ func (container *Container) Start() (err error) {
 			container.toDisk()
 			container.cleanup()
 			container.LogEvent("die")
+			container.LogEvent("exit")
```
Comment (duglin) [inline daemon/container.go]: From looking at the code, it appears that "die" is the last event we get for a container this is being shutdown/stopped.  If we're going to have a new even that indicates something wrong happened (or at least not a normal exit) then I would prefer to see that event **before** the "die", not after.

Comment (duglin) [inline daemon/container.go]: Let me elaborate, I think its important to has consistent "book-ends" around the events for a container: 'create' at the start, and 'die' at the end.  This allows for a consistent processing of the events stream - the user can be assured that once they see 'die' they can stop looking for events for that container.  With this code they now have to wait (for a racey amount of time) to see if 'exit' shows up later on.  It makes it much more complicated w.r.t. the event processing.

**Comment by mgood** [comment]
@duglin yes, I would probably agree that "die" should come last if starting over, but since "kill" and "stop" already occur after the "die" event, it seemed more consistent to me to also include "exit" after the "die" event. The docs contain a diagram [here under "Docker Events"](https://docs.docker.com/reference/api/docker_remote_api/) showing that existing behavior.

I agree that the raciness of checking for events after "die" can be an issue, which happens now with checking for "kill" or "stop" which occur after "die". I discussed this in more detail in my comment from Apr 23, but this "exit" event is meant to address that kind of racy behavior checking for "kill" or "stop". Since this "exit" event would occur whenever the container exited for a reason other than "kill" or "stop", you can distinguish those 3 cases without checking for the "die" event itself. Listeners that only care about 

It looks like with 8232312c1e705753d3db82dca3d9bb23e59c3b52 when you were cleaning up the LogEvent calls, by moving the "kill" event, it changed the event behavior a bit as well.

The events for "docker kill" changed from `die, kill` to `kill, die`
and "docker stop" changed from `die, stop` to `kill, die, stop`.

This PR includes some new tests from "kill" and "stop" based on the previous behavior, which is why they failed in CI. I'd like to update this PR to restore the previous event flow so that it remains consistent with the existing behavior which is now documented in the "Docker Events" diagram I mentioned above.

**Comment by duglin** [comment]
@mgood I'd suggest that we keep "docker kill" as is, it makes sense that kill comes before die since (to me) kill means "we sent it a kill signal" and then the container died - which is what happened.
I think we should try to fix "docker stop" to be: stop, kill, die since (again) that's what is really happen - we asked for it to stop (by sending it a kill), and then it died. An argument can be made for not even including 'stop' since its not really an action by itself - IOW, a CLI "stop" is really a daemon "kill" - and I'd actually prefer that.  Why have two events for one action taken on the container?  Might make sense in some cases, but not this.  And then fix the docs/picture.
@LK4D4 what do you think?

**Comment by duglin** [comment]
@mgood slight amendment... I guess keeping the stop in there isn't a bad idea if people want to know that the cli used stop() instead of kill(), which I guess could be useful, but even then I think putting it before the "kill" in the flow is ok

**Comment by mgood** [comment]
@duglin I see the logic behind your suggestions, and I think they might be reasonable if this was earlier in the design process, but introducing those changes now can break existing apps that rely on the events API.

We rely on the events API heavily for Gliderlabs projects like registrator and logspout, and expect these behaviors to be pretty consistent across Docker versions. Removing or moving around existing events would probably break some existing apps we've written. For example moving the "stop" event before "die" changes the meaning of the event quite significantly. We expect "stop" to mean that the container has already been stopped, not that it will be stopped soon. When we get a "stop" event, we may check the exit status, which might not be set yet if the container is still running. Making this work across Docker versions would require some complicated workarounds.

The same applies to "kill". We also expect "kill" to mean that it completed the "kill" API call, not that it's just sent the signal. The difference is subtle, but also has the potential to break things.

The docs have so far been a bit ambiguous on the meaning of the events. I worked on the initial version of the diagram that now appears in the docs because it was difficult to know how to program against the events API without more definitive knowledge of how and when the events would be triggered.

I think you raise some interesting points, but I'd prefer to see some of those addressed in v2 of the API instead of making potentially breaking changes to v1.

**Comment by cpuguy83** [comment]
@mgood There are some test failures related to events.

**Comment by mgood** [comment]
@cpuguy83 yeah, those should be fixed now. I included some extra tests for "kill" and "stop" which failed due to an intervening change to the location of the "kill" event. I've included a fix now to restore the previous behavior for the "kill" event.

**Comment by icecrime** [comment]
Ping @cpuguy83!

**Comment by icecrime** [comment]
Also ping @RichardScothern if you can please give this a review/test!

**Comment by tiborvass** [comment]
How necessary is this if we have something like #14012 ?

**Comment by progrium** [comment]
@tiborvass this is out of band from the cli ... since events are just notifications without much payload consumed via the api. you'd then inspect containers to get more data, like what the exit code was. so i don't believe there is much interaction with #14012

**Comment by mgood** [comment]
@tiborvass that PR looks nice, but I don't think it helps here. Those new exit codes would help distinguish the reason a container failed to start, but with this I'm interested in distinguishing when a long-running container exits unexpectedly.

File: daemon/container.go (hunk 1 above)
Comment (jfrazelle) [inline daemon/container.go]: I agree

File: daemon/monitor.go [hunk 2]
```diff
@@ -179,6 +182,9 @@ func (m *containerMonitor) Start() error {
 			m.container.LogEvent("oom")
 		}
 		m.container.LogEvent("die")
+		if !m.shouldStop {
+			m.container.LogEvent("exit")
```
Comment (jfrazelle) [inline daemon/monitor.go]: i agree with @duglin this should come _before_ die

**Comment by LK4D4** [comment]
@mgood @jfrazelle @duglin I have no preference about order of events. But we need to move this PR forward, so if @jfrazelle and @duglin think that exit should be before die, then probably it should.

**Comment by icecrime** [comment]
@LK4D4 I 100% agree. If that is the only blocker, we should consider carrying as the author has waited long enough.

**Comment by calavera** [comment]
Carrying this PR in https://github.com/docker/docker/pull/15669
